### PR TITLE
[python] added f-strings to helpers/parameter_generator.py

### PR DESCRIPTION
**Pull request by NovusEdge**

### CODE REVIEW

**Comment by NovusEdge** [comment]
#4136

File: helpers/parameter_generator.py [hunk 1]
```diff
@@ -171,20 +171,20 @@ def set_one_var_from_string(name, param_type, checks):
...
         if len(checks) > 0:
             check_mapper = {"<": "LT", ">": "GT", "<=": "LE", ">=": "GE"}
             for check in checks:
                 value, sign = parse_check(check)
-                ret += "  CHECK_%s(%s, %s);\n" % (check_mapper[sign], name, value)
+                ret += f"  CHECK_{check_mapper[sign]}({name}, {value});\n"
         ret += "\n"
     else:
-        ret += "  if (GetString(params, \"%s\", &tmp_str)) {\n" % (name)
+        ret += f"  if (GetString(params, \"{name}\", &tmp_str)) {\n"
```
Comment (NovusEdge) [inline helpers/parameter_generator.py:182]: Note to self: fix this in the next commit

Comment (jameslamb) [reply helpers/parameter_generator.py:182]: @NovusEdge are you still interested in this PR? Do you need any help?

Comment (NovusEdge) [reply helpers/parameter_generator.py:182]: > @NovusEdge are you still interested in this PR?
I am, but I keep running into issues with the linting 😓

Comment (jameslamb) [reply helpers/parameter_generator.py:182]: ok what issues are you facing and what questions can we answer?

Comment (NovusEdge) [reply helpers/parameter_generator.py:182]: [this](https://github.com/microsoft/LightGBM/pull/4146/checks?check_run_id=2229633254#step:3:533)

Comment (jameslamb) [reply helpers/parameter_generator.py:182]: What specific question to you have about those checks?

Have you tried reproducing them with the steps in https://github.com/microsoft/LightGBM/pull/4142#pullrequestreview-624538246?

**Comment by jameslamb** [comment]
I merged in the latest changes from the `master` branch to this PR tonight, following the steps described in https://github.com/microsoft/LightGBM/pull/4144#pullrequestreview-656153808.

I also patched a few small mistakes in https://github.com/microsoft/LightGBM/pull/4146/commits/63ebdaf58f0d88171d69a7a0c397acc63c095b78.

Importantly, the linting errors observed in https://github.com/microsoft/LightGBM/pull/4146#discussion_r604577999 were a result of incorrect escaping of literal curly braces. I learned tonight, from https://docs.python.org/3/library/string.html#format-string-syntax, that you should use a `{{` to indicate a literal `{` (and `}}` for a literal `}`) in an f-string.

```python
x = "hello"
f"{x}, {{curly braces}}"

# 'hello, {curly braces}'
```

I also ran `python helpers/parameter_generator.py` to be sure that the changes don't change the output produced by that script.

@StrikerRUS I think this is now ready for review.

**Review by jameslamb — COMMENTED** [review]
These changes look good to me, but since I added a few changes I don't think my approval should count towards a merge.

File: helpers/parameter_generator.py [hunk 2]
```diff
@@ -205,34 +205,36 @@ def gen_parameter_description(sections, descriptions, params_rst):
...
             elif checks_len == 1:
                 number, sign = parse_check(checks[0])
-                checks_str = ', constraints: ``{0} {1} {2}``'.format(name, sign, number)
+                checks_str = f', constraints: ``{name} {sign} {number}``'
             else:
                 checks_str = ''
-            main_desc = '-  ``{0}`` :raw-html:`<a id="{0}" title="Permalink to this parameter" href="#{0}">&#x1F517;&#xFE0E;</a>`, default = ``{1}``, type = {2}{3}{4}{5}'.format(name, default, param_type, options_str, aliases_str, checks_str)
+            main_desc = f'-  ``{name}`` :raw-html:`<a id="{name}" title="Permalink to this parameter" href="#{name}">&#x1F517;&#xFE0E;</a>`, default = ``{default}``, type = {param_type}{options_str}{aliases_str}{checks_str}'
             params_to_write.append(main_desc)
             params_to_write.extend([' ' * 3 * int(desc[0][-1]) + '-  ' + desc[1] for desc in param_desc['desc']])
```
Comment (StrikerRUS) [inline helpers/parameter_generator.py:239]:
```suggestion
            params_to_write.extend([f"{' ' * 3 * int(desc[0][-1])}-  {desc[1]}" for desc in param_desc['desc']])
```

File: helpers/parameter_generator.py [hunk 3]
```diff
@@ -281,7 +283,7 @@ def gen_parameter_code(config_hpp, config_out_cpp):
     str_to_write += "  static std::unordered_map<std::string, std::string> aliases({\n"
 
     for pair in alias:
-        str_to_write += "  {\"%s\", \"%s\"},\n" % (pair[0], pair[1])
+        str_to_write += f"  {{\"{pair[0]}\", \"{pair[1]}\"}},\n"
```
Comment (StrikerRUS) [inline helpers/parameter_generator.py:286]:
```suggestion
        str_to_write += f'  {{"{pair[0]}", "{pair[1]}"}},\n'
```

File: helpers/parameter_generator.py [hunk 4]
```diff
@@ -291,7 +293,7 @@ def gen_parameter_code(config_hpp, config_out_cpp):
     str_to_write += "  static std::unordered_set<std::string> params({\n"
 
     for name in names:
-        str_to_write += "  \"%s\",\n" % (name)
+        str_to_write += f"  \"{name}\",\n"
```
Comment (StrikerRUS) [inline helpers/parameter_generator.py:296]:
```suggestion
        str_to_write += f'  "{name}",\n'
```

File: helpers/parameter_generator.py [hunk 5]
```diff
@@ -321,11 +323,11 @@ def gen_parameter_code(config_hpp, config_out_cpp):
             name = y["name"][0]
             if "vector" in param_type:
                 if "int8" in param_type:
-                    str_to_write += "  str_buf << \"[%s: \" << Common::Join(Common::ArrayCast<int8_t, int>(%s), \",\") << \"]\\n\";\n" % (name, name)
+                    str_to_write += f"  str_buf << \"[{name}: \" << Common::Join(Common::ArrayCast<int8_t, int>({name}), \",\") << \"]\\n\";\n"
```
Comment (StrikerRUS) [inline helpers/parameter_generator.py:326]:
```suggestion
                    str_to_write += f'  str_buf << "[{name}: " << Common::Join(Common::ArrayCast<int8_t, int>({name}), ",") << "]\\n";\n'
```

File: helpers/parameter_generator.py [hunk 6]
```diff
@@ -321,11 +323,11 @@ def gen_parameter_code(config_hpp, config_out_cpp):
...
                 else:
-                    str_to_write += "  str_buf << \"[%s: \" << Common::Join(%s, \",\") << \"]\\n\";\n" % (name, name)
+                    str_to_write += f"  str_buf << \"[{name}: \" << Common::Join({name}, \",\") << \"]\\n\";\n"
```
Comment (StrikerRUS) [inline helpers/parameter_generator.py:328]:
```suggestion
                    str_to_write += f'  str_buf << "[{name}: " << Common::Join({name}, ",") << "]\\n";\n'
```

File: helpers/parameter_generator.py [hunk 7]
```diff
@@ -321,11 +323,11 @@ def gen_parameter_code(config_hpp, config_out_cpp):
...
             else:
-                str_to_write += "  str_buf << \"[%s: \" << %s << \"]\\n\";\n" % (name, name)
+                str_to_write += f"  str_buf << \"[{name}: \" << {name} << \"]\\n\";\n"
```
Comment (StrikerRUS) [inline helpers/parameter_generator.py:330]:
```suggestion
                str_to_write += f'  str_buf << "[{name}: " << {name} << "]\\n";\n'
```

**Review by StrikerRUS — CHANGES_REQUESTED** [review]

**Comment by StrikerRUS** [comment]
Ah, missed one suggestion:
https://github.com/microsoft/LightGBM/blob/08d1ce4bbf9391f5cc945f5f015840f4a57774a6/helpers/parameter_generator.py#L300
This string is not needed to be converted into f-string. But I think we can change outer `"` quotes to `'` ones to remove escaping in this PR.

**Comment by jameslamb** [comment]
> Ah, missed one suggestion:
> https://github.com/microsoft/LightGBM/blob/08d1ce4bbf9391f5cc945f5f015840f4a57774a6/helpers/parameter_generator.py#L300
> 
> 
> This string is not needed to be converted into f-string. But I think we can change outer `"` quotes to `'` ones to remove escaping in this PR.

ok. Updated in https://github.com/microsoft/LightGBM/pull/4146/commits/62f3380a4cd55c6d6deb908e258accd39dd7468c

**Review by StrikerRUS — APPROVED** [review]

**Pull request merged**